File: web/src/api/post.ts
import type {
  Category,
  CreateCategoryReq,
  CreateLabelReq,
  CreatePostReq,
  Label,
  Post,
  PostLikeResponse,
  PostListRequest,
  PostListResponse,
  UpdateCategoryReq,
  UpdateLabelReq,
  UpdatePostReq,
} from '@/models/post'
import type { Resp } from '@/models/resp'
import { axiosClient } from '@/lib/api/client'
import { unwrapResponse } from '@/lib/api/response'

export async function getPost(slugOrID: string): Promise<Post> {
  const response = await axiosClient.get<Resp<Post>>(
    `/post/p/${encodeURIComponent(slugOrID)}`,
  )
  return unwrapResponse(response.data)
}

export async function listPosts(req: PostListRequest = {}): Promise<PostListResponse> {
  const response = await axiosClient.get<Resp<PostListResponse>>(
    '/post/list',
    {
      params: req,
    },
  )
  return unwrapResponse(response.data)
}

export async function getRandomPost(): Promise<Post> {
  const response = await axiosClient.get<Resp<Post>>(
    '/post/random',
  )

  return unwrapResponse(response.data)
}

export async function createPost(req: CreatePostReq): Promise<Post> {
  const response = await axiosClient.post<Resp<Post>>(
    '/post/p',
    req,
  )
  return unwrapResponse(response.data)
}

export async function updatePost(
  id: number,
  req: UpdatePostReq,
): Promise<Post> {
  const response = await axiosClient.put<Resp<Post>>(
    `/post/p/${id}`,
    req,
  )
  return unwrapResponse(response.data)
}

export async function deletePost(id: number): Promise<void> {
  await axiosClient.delete(`/post/p/${id}`)
}

export async function togglePostLike(id: number): Promise<PostLikeResponse> {
  const response = await axiosClient.post<Resp<PostLikeResponse>>(
    `/post/p/${id}/like`,
  )

  return unwrapResponse(response.data)
}

export async function createCategory(
  req: CreateCategoryReq,
): Promise<Category> {
  const response = await axiosClient.post<Resp<Category>>(
    '/post/c',
    req,
  )

  return unwrapResponse(response.data)
}

export async function updateCategory(
  id: number,
  req: UpdateCategoryReq,
): Promise<Category> {
  const response = await axiosClient.put<Resp<Category>>(
    `/post/c/${id}`,
    req,
  )

  return unwrapResponse(response.data)
}

export async function deleteCategory(id: number): Promise<void> {
  await axiosClient.delete(`/post/c/${id}`)
}

export async function createLabel(req: CreateLabelReq): Promise<Label> {
  const response = await axiosClient.post<Resp<Label>>(
    '/post/l',
    req,
  )

  return unwrapResponse(response.data)
}

export async function updateLabel(
  id: number,
  req: UpdateLabelReq,
): Promise<Label> {
  const response = await axiosClient.put<Resp<Label>>(
    `/post/l/${id}`,
    req,
  )

  return unwrapResponse(response.data)
}

export async function deleteLabel(id: number): Promise<void> {
  await axiosClient.delete(`/post/l/${id}`)
}

File: internal/router/apiv1/post.go
package apiv1

import (
	"github.com/cloudwego/hertz/pkg/route"
	"github.com/zyj/my-blog/internal/controller"
	"github.com/zyj/my-blog/internal/middleware"
	"github.com/zyj/my-blog/pkg/constant"
)

func registerPostRoutes(group *route.RouterGroup) {
	post := group.Group("/post")

	post.GET(
		"/p/:slug_or_id",
		middleware.UseAuth(false),
		controller.GetPostDetail,
	)
	post.GET(
		"/list",
		middleware.UseAuth(false),
		controller.ListPosts,
	)
	post.GET(
		"/categories",
		middleware.UseAuth(false),
		controller.ListCategories,
	)
	post.GET(
		"/labels",
		middleware.UseAuth(false),
		controller.ListLabels,
	)
	post.GET(
		"/random",
		middleware.UseAuth(false),
		controller.GetRandomPost,
	)
	post.POST(
		"/p/:id/like",
		middleware.UseAuth(true),
		controller.TogglePostLike,
	)

	admin := group.Group(
		"/post",
		middleware.UseAuth(true),
		middleware.UseRole(constant.RoleAdmin),
	)

	admin.POST("/p", controller.CreatePost)
	admin.PUT("/p/:id", controller.UpdatePost)
	admin.DELETE("/p/:id", controller.DeletePost)

	admin.POST("/c", controller.CreateCategory)
	admin.PUT("/c/:id", controller.UpdateCategory)
	admin.DELETE("/c/:id", controller.DeleteCategory)

	admin.POST("/l", controller.CreateLabel)
	admin.PUT("/l/:id", controller.UpdateLabel)
	admin.DELETE("/l/:id", controller.DeleteLabel)
}
